cmd/juju/application/deployer/bundlehandler.go

```diff
@@ -1439,8 +1437,196 @@ func resolve(placeholder string, results map[string]string) string {
+			Endpoint2: relation.Endpoints[1].Name,
+		})
+	}
+	// Get all the annotations.
+	annotations, err := apiRoot.GetAnnotations(annotationTags)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for _, result := range annotations {
+		if result.Error.Error != nil {
+			return nil, errors.Trace(result.Error.Error)
+		}
+		tag, err := names.ParseTag(result.EntityTag)
+		if err != nil {
+			return nil, errors.Trace(err) // This should never happen.
+		}
+		switch kind := tag.Kind(); kind {
+		case names.ApplicationTagKind:
+			mod.Applications[tag.Id()].Annotations = result.Annotations
+		case names.MachineTagKind:
+			mod.Machines[tag.Id()].Annotations = result.Annotations
+		default:
+			return nil, errors.Errorf("unexpected tag kind for annotations: %q", kind)
+		}
+	}
+	// Add in the model sequences.
+	sequences, err := apiRoot.Sequences()
+	if err == nil {
+		mod.Sequence = sequences
+	} else if !errors.IsNotSupported(err) {
+		return nil, errors.Annotate(err, "getting model sequences")
+	}
+
+	// When dealing with bundles the current model generation is always used.
+	configValues, err := apiRoot.GetConfig(model.GenerationMaster, appNames...)
+	if err != nil {
+		return nil, errors.Annotate(err, "getting application options")
+	}
+	for i, cfg := range configValues {
+		options := make(map[string]interface{})
+		// The config map has values that looks like this:
+		//  map[string]interface {}{
+		//        "value":       "",
+		//        "source":     "user", // or "unset" or "default"
+		//        "description": "Where to gather metrics from.\nExamples:\n  host1.maas:9090\n  host1.maas:9090, host2.maas:9090\n",
+		//        "type":        "string",
+		//    },
+		// We want the value iff default is false.
+		for key, valueMap := range cfg {
+			value, err := applicationConfigValue(key, valueMap)
+			if err != nil {
+				return nil, errors.Annotatef(err, "bad application config for %q", appNames[i])
+			}
+			if value != nil {
+				options[key] = value
+			}
+		}
+		mod.Applications[appNames[i]].Options = options
+	}
+	// Lastly get all the application constraints.
+	constraintValues, err := apiRoot.GetConstraints(principalApps...)
+	if err != nil {
+		return nil, errors.Annotate(err, "getting application constraints")
+	}
+	for i, value := range constraintValues {
+		mod.Applications[principalApps[i]].Constraints = value.String()
+	}
+
+	mod.ConstraintsEqual = func(a, b string) bool {
+		// Since the constraints have already been validated, we don't
+		// even bother checking the error response here.
+		ac, _ := constraints.Parse(a)
+		bc, _ := constraints.Parse(b)
+		return reflect.DeepEqual(ac, bc)
+	}
+
+	return mod, nil
+}
+
+// applicationConfigValue returns the value if it is not a default value.
+// If the value is a default value, nil is returned.
+// If there was issue determining the type or value, an error is returned.
+func applicationConfigValue(key string, valueMap interface{}) (interface{}, error) {
```

Comment:
This got changed due to the merge conflict. Not sure if this got added in 2.9 or it is something that was removed with the deploy process overhaul.

Reply:
applicationConfigValue is now part of the bundle package: https://github.com/juju/juju/blob/f76391aee8f3ce5eb07eaddd8397464f5d48f3e8/cmd/juju/application/bundle/bundle.go#L190